_posts/2021-04-05-fmf-unified-testing.md

```diff
@@ -0,0 +1,156 @@
+
+## Recent Fedora CI changes
+
+As mentioned above, Fedora's gating tests recently switched to run on the very same [testing-farm](https://testing-farm.gitlab.io/api/), so that Fedora gating and Packit tests run in the same environment and thus compare very well. It also supports FMF/TMT test metadata in addition to the old STI.
+
+This gets us *really* close to the goal of sharing the tests upstream and downstream. The only bit that's missing is that there is no clean way how to run tests that are contained in your upstream tarball -- right now, the packaging dist-git must have a top-level [FMF test plan](https://src.fedoraproject.org/rpms/cockpit/blob/rawhide/f/plans/upstream.fmf) like this:
+
+```yaml
+discover:
+  how: fmf
+  repository: https://github.com/cockpit-project/cockpit
+  # FIXME: get rid of the hardcoding: https://github.com/psss/tmt/issues/585
+  ref: "241"
+execute:
+  how: tmt
+```
+
+This checks out the tests from your upstream project git. This is awkward, as it requires accessing a remote git even though the tests are already in the source tarball; and it needs to know which exact tag to check out, so that the tests match your packaged release. This requires some [tmt design discussion](https://github.com/psss/tmt/issues/585), but for now we just [hacked our release scripts](https://github.com/cockpit-project/cockpituous/commit/2ef3f6c99912) to automatically bump the `ref:` in the test plan when it commits the new release to dist-git.
+
+Even with this hack, cockpit's [commit to move from STI to upstream FMF](https://src.fedoraproject.org/rpms/cockpit/c/0fee2830080033ea2be13be30d156b51dcf75b7d) was still a major net gain: The added files are rather static and uninteresting, and all the test setup and the tests themselves run straight from upstream now.
+
+## Putting it all together
+
+Cockpit's [starter-kit](https://github.com/cockpit-project/starter-kit/#running-tests-in-ci), the basis for creating your own Cockpit UIs, contains all of this now: FMF metadata, [setup scripts](https://github.com/cockpit-project/starter-kit/tree/master/test/browser), [packit.yaml](https://github.com/cockpit-project/starter-kit/blob/master/packit.yaml), and [documentation](https://github.com/cockpit-project/starter-kit#running-tests-in-ci).
+
+[Doing this for cockpit](https://github.com/cockpit-project/cockpit/pull/15504) was a little more involved, mostly because packit's `create-archive` step is currently rather limited -- it needs to work in a 768 MiB VM and finish in 30 minutes, which is just not enough for webpack. So we [build the tarballs](https://github.com/cockpit-project/cockpit/blob/master/.github/workflows/build-dist.yml) in a GitHub workflow and just download these from packit (we want to do that anyway, as these are highly useful for speeding up reviews and local development as well).
```

Comment:
Is this Cockpit  specific? If so, will every Cockpit based package run into this? If not, does every package need to figure this out for themeselves?

Reply:
Not every package, e.g. starter-kit and cockpit-podman don't need this. Pretty much every project *except* large webpack ones (i.e. 95% of Fedora packages) should not need this hack. I clarified this in this paragraph now, I hope.

Reply:
> Is this Cockpit specific?

Yes, because cockpit has higher requirements to create an archive from a checkout (aside from just running `git archive`).

> If so, will every Cockpit based package run into this? If not, does every package need to figure this out for themeselves?

Martin managed to assemble archives for some cockpit's subprojects within the packit's constrained quota even though he run into several issues which we tried to address promptly.

Overall, this step is really tricky since for some upstream projects it's trivial and a no-brainer, but when archive creation is complex, it takes non-trivial effort to make it work in packit.

At this point I really hope that cockpit-based projects could already utilize a boilerplate template for .packit.yaml and share it.

Reply:
Right, me too -- we are still making heavy changes to the "pre-built dist/" stuff in cockpit, cockpit-machines and such. These are useful in so many ways other than packit, but we are still working out their storage and how to weave them into PRs, local testing, release process, convenient CLI and such. As such, I'd rather not advertise that as a finished solution yet.